numpyro/distributions/constraints.py

```diff
@@ -412,10 +424,47 @@ def log_abs_det_jacobian(self, x, y, intermediates=None):
         return np.full(np.shape(x) if self.event_dim == 0 else np.shape(x)[:-1], 0.)
 
 
+class InvCholeskyTransform(Transform):
+    r"""
+    Transform via the mapping :math:`y = x @ x.T`, where `x` is a lower
+    triangular matrix with positive diagonal.
+    """
+    event_dim = 2
+
+    def __init__(self, domain=lower_cholesky):
+        assert domain in [lower_cholesky, corr_cholesky]
+        self.domain = domain
+
+    @property
+    def codomain(self):
+        if self.domain is lower_cholesky:
+            return positive_definite
+        elif self.domain:
+            return corr_matrix
+
+    def __call__(self, x):
+        return np.matmul(x, np.swapaxes(x, -2, -1))
+
+    def inv(self, y):
+        return np.linalg.cholesky(y)
+
+    def log_abs_det_jacobian(self, x, y, intermediates=None):
+        if self.domain is lower_cholesky:
+            # Ref: http://web.mit.edu/18.325/www/handouts/handout2.pdf page 13
+            n = np.shape(x)[-1]
+            order = np.arange(n, 0, -1)
+            return n * np.log(2) + np.sum(order * np.log(np.diagonal(x, axis1=-2, axis2=-1)), axis=-1)
+        else:
+            # NB: see derivation in LKJCholesky implementation
+            n = np.shape(x)[-1]
+            order = np.arange(n - 1, -1, -1)
+            return np.sum(order * np.log(np.diagonal(x, axis1=-2, axis2=-1)), axis=-1)
+
+
 class LowerCholeskyTransform(Transform):
     domain = real_vector
     codomain = lower_cholesky
-    event_dim = 1
+    event_dim = 2
```

Comment:
I changed `event_dim` to 2 to:
+ match Pyro implementation
+ make log_abs_det_transform of ComposeTransform([LowerCholesky, InvCholesky]) correct: we need to keep both transforms have the same event_dim
+ make [event_shape](https://github.com/pyro-ppl/numpyro/blob/master/numpyro/distributions/distribution.py#L272) of TransformedDistribution(dist.Normal(np.zeros(3), 1), LowerCholeskyTransform) correct.

Side effect:
+ ComposedTransformed([AffineTransform(domain=real_vector), LowerCholeskyTransform]) will have wrong log_det due to [this line](https://github.com/pyro-ppl/numpyro/blob/master/numpyro/distributions/constraints.py#L288) (the former part has event_dim=1, while the later part has event_dim)

To fix the above site effect, we can set `event_dim=1` for both LowerCholeskyTransform and InvCholeskyTransform. In that case, shape sematic of TransformedDistribution(dist.Normal(np.zeros(3), 1), LowerCholeskyTransform) will be wrong. However, I think that no user wants to construct those kinds of ComposeTransform, TransformedDistribution so either way (setting `event_dim=1` or `event_dim=2` is fine to me). Personally, I prefer to set `event_dim=1` because it is helpful to think the codomain lower_cholesky as a vector, rather than a matrix (when computing jacobian), but I don't want to break the convention in Pyro/PyTorch or create confusion for users. So I set `event_dim=2` here.

Reply:
I think this is fine. I will need to think about it more to see if there is a more general way to conceptualize event dimensions in transforms, but let us spend time on this when we have more uses in hand.